### ISSUE TITLE
Training auto-sklearn models on a dask cluster

### ISSUE THREAD

**Issue by bmmalone** [automl/auto-sklearn#351]
Hi,

I have been trying to submit training calls for auto-sklearn to a [dask cluster](https://dask.pydata.org/en/latest/) and was running into trouble.

The problem ends up being that the call to `run` in `evaluation/__init__.py` passes a logger to pynisher (it creates the logger on Line ~157), which then includes the logger when it wraps the call into SMAC; the logger includes an un-picklable RLock. However, dask pickles the function calls and all arguments to distribute them across the cluster. Thus, it fails when trying to pickle the logger. (Somewhat surprisingly to me, the multiprocessing.Queue is fine.)

I updated the code in `evaluation/__init__.py` to not pass a logger, as well as the code in pynisher to check if it has a logger before logging, and then all is well; dask pickles the calls to SMAC and distributes them as expected.

Below, I posted a small example that does an outer cross-validation loop based on `joblib.Parallel` with a dask backend that causes the error. When I use the updated "no logger" code, it behaves as expected.

Right now, the code I have is a bit hacky (it just never passes a logger to pynisher), though I believe this could be added as an option to the AutoSklearnClassifier and Regressor constructors.

Please let me know if there would be a good way to incorporate what I already have (via PRs, etc.), or if I can clarify anything so far. Thanks.

Have a good day,
Brandon

Before running the script, the dask cluster can be started using the [`dash-ssh`](http://distributed.readthedocs.io/en/latest/setup.html#using-ssh) command, and the `cluster_location` below is the scheduler address from that. 

```
# connect to the dask cluster
from dask.distributed import Client as DaskClient

cluster_location = '192.168.0.81:8786'
client = DaskClient(address=cluster_location)
client.restart() # for testing, make sure we clear caches, etc.

import joblib
import distributed.joblib

import autosklearn.classification
import sklearn.datasets
import sklearn.model_selection

def _outer_cv(fold, num_folds):
    classifier = autosklearn.classification.AutoSklearnClassifier(time_left_for_this_task=30)
    
    X, y = sklearn.datasets.load_iris(return_X_y=True)
    cv = sklearn.model_selection.KFold(num_folds)
    
    splits = cv.split(X)
    train, test = next(splits) # would really use more_itertools.nth
    
    X_train, y_train = X[train], y[train]
    X_test, y_test = X[test], y[test]
    
    classifier_fit = classifier.fit(X_train, y_train)
    y_pred =classifier.predict(X_test)
    
    return (y_test, y_pred)

num_folds = 5
folds = range(num_folds)

with joblib.parallel_backend("dask.distributed", scheduler_host=cluster_location):
    ys = joblib.Parallel()(joblib.delayed(_outer_cv)(fold, num_folds) for fold in folds)
```

**Comment by mfeurer** [comment]
Thanks for your offer to contribute. I'm out of office until the end of September and can only sporadically answer emails, but not take care of larger requests. I'll have a look at this when I'm back in the office.

**Comment by bmmalone** [comment]
Hi Matthias,

Okay. The fixes are relatively small, but they are pretty deep in the codebase, so it would definitely be good for someone more familiar with the internals than me to ensure they don't have any unintended consequences. Please just let me know if I can help, provide more details, etc. I hope you enjoy your vacation/travel.

Have a good day,
Brandon

**Comment by etemiz** [comment]
Hi @bmmalone, 
Thank for sharing your experience.
Could you add the code please?

**Comment by bmmalone** [comment]
Hi @etemiz,

Unfortunately, I cannot (for now) share the code due to work constraints. (We hope to resolve this soon.)

Still, the basic idea is as described above; in the `run` function in `evaluation/__init__.py`, do not pass logger. Then, in `pynisher/limit_function_call.py`, just ignore the logger if it is not given.

Finally, it is also necessary to say that the current process is not a daemon, or otherwise dask is unable to spawn processes and fails. I added this to `limit_function_call.py` around line 215:

```
multiprocessing.current_process()._config['daemon'] = False
```

I do not really know if changing the daemon status has other implications, but it seems to work.

---

Also, just to clarify, this works for distributing multiple instances of auto-sklearn (such as when performing outer cross-validation); it does not distribute a single instance of auto-sklearn across the cluster.

**Comment by mfeurer** [comment]
This will be fixed in #946  As a bonus, #946 also adds running the target function evaluations themselves in parallel using Dask.
